File: Source/Example.Streams.Chat.Client/ChatClient.cs
using System;
using System.Threading.Tasks;

using Orleankka;

namespace Example
{
    class ChatClient
    {
        readonly ActorRef<IChatUser> user;
        readonly IActorSystem system;
        StreamSubscription subscription;

        public ChatClient(IActorSystem system, string user, string room)
        {
            this.user = system.TypedActorOf<IChatUser>(user);
            this.system = system;
            RoomName = room;
        }

        public async Task Join()
        {
            await Subscribe();
            await user.Tell(new Join {Room = RoomName});
        }

        async Task Subscribe()
        {
            var room = system.StreamOf("sms", RoomName);
            subscription = await room.Subscribe<ChatRoomMessage>(message =>
            {
                if (message.User != UserName)
                    Console.WriteLine(message.Text);
            });
        }

        public Task Resubscribe() => Subscribe();

        public async Task Leave()
        {
            await subscription.Unsubscribe();
            await user.Tell(new Leave {Room = RoomName});
        }

        public async Task Say(string message)
        {
            await user.Tell(new Say {Room = RoomName, Message = message});
        }

        string UserName => user.Path.Id;
        string RoomName { get; }
    }
}

File: Source/FSharp.Example.Streams.Chat.Client/ChatClient.fs
module Client.ChatClient

open Orleankka
open Orleankka.FSharp
open Messages

type ChatClient = {
   UserName: string
   User: ActorRef<obj>
   RoomName: string   
   Room: StreamRef<ChatRoomMessage>   
   Subscription: Option<StreamSubscription>
}

let join (client:ChatClient) = task {   
   let! sb = client.Room.Subscribe (fun messge ->
      if messge.UserName <> client.UserName then printfn "%s" messge.Text
   )

   do! client.User <! Join(client.RoomName)

   return { client with Subscription = Some sb }
}

let leave (client:ChatClient) = task {
   do! client.Subscription.Value.Unsubscribe() |> Task.awaitTask
   do! client.User <! Leave(client.RoomName)
}

let say (client:ChatClient) (message:string) = task {
   do! client.User <! Say(client.RoomName, message)   
}